packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts

```diff
@@ -0,0 +1,130 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryProperties } from "@fluidframework/common-definitions";
+import { ICriticalContainerError } from "@fluidframework/container-definitions";
+import {
+    IConfigProvider,
+    IFluidErrorBase,
+    LoggingError,
+    MonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { oneDayMs } from "./garbageCollection";
+
+const sweepReadyUsageDetectionKey = "Fluid.GarbageCollection.Dogfood.SweepReadyUsageDetection";
+const blackoutPeriodDaysKey = "Fluid.GarbageCollection.Dogfood.SweepReadyUsageDetection.BlackoutPeriodDays";
+const closuresStorageKey = "Fluid.GarbageCollection.Dogfood.SweepReadyUsageDetection.Closures";
+
+/**
+ * Feature gate key to enable closing the container if SweepReady objects are used.
+ * Value should contain keywords "mainContainer" and/or "summarizer" to enable detection in each container type
+ */
+const sweepReadyUsageDetectionSetting = {
+    read(config: IConfigProvider) {
+        const value = config.getString(sweepReadyUsageDetectionKey);
+        if (value === undefined) {
+            return { mainContainer: false, summarizer: false };
+        }
+        return {
+            mainContainer: value.indexOf("mainContainer") >= 0,
+            summarizer: value.indexOf("summarizer") >= 0,
+        };
+    },
+};
+
+/**
+ * Error class raised when a SweepReady object is used, indicating a bug in how
+ * references are managed in the container by the application, or a bug in how
+ * GC tracks those references.
+ *
+ * There's a chance for false positives when this error is raised by a Main Container,
+ * since only the Summarizer has the latest truth about unreferenced node tracking
+ */
+export class SweepReadyUsageError extends LoggingError implements IFluidErrorBase {
+    /** This errorType will be in temporary use (until Sweep is fully implemented) so don't add to any errorType type */
+    public errorType: string = "unreferencedObjectUsedAfterGarbageCollected";
+}
+
+/**
+ * This class encapsulates the logic around what to do when a SweepReady object is used.
+ * There are several tactics we plan to use in Dogfood environments to aid diagnosis of these cases:
+ *  - Closing the container when either the main or summarizer client detects this kind of violation
+ *  - Throttling the frequency of these crashes via a "Blackout Period" per container per device
+ */
+export class SweepReadyUsageDetectionHandler {
+    private readonly localStorage: Pick<Storage, "getItem" | "setItem">;
+
+    constructor(
+        /** Unique key for this container for diagnostic purposes */
+        private readonly uniqueContainerKey: string,
+        private readonly mc: MonitoringContext,
+        private readonly closeFn: (error?: ICriticalContainerError) => void,
+        localStorageImpl?: Pick<Storage, "getItem" | "setItem">,
```

Comment:
Why aren't we setting this in the main code path?

Reply:
We'll use browser's localStorage (see 71: `this.localStorage = localStorage ?? noopStorage`).

I could call it override instead of impl